File: objective-c/binary-search/bs/Tests/BinarySearch.m
#import "BinarySearch.h"
@implementation BinarySearch
- (instancetype)initWithArray:(NSArray *)lst {
    for (int i = 1; i < lst.count; ++i) {
        if ([lst[i] integerValue] < [lst[i - 1] integerValue]) { return nil; }
    }
    _list = lst;
    return self;
}
- (NSInteger)searchFor:(NSInteger)search {
    for (NSInteger lo = 0, hi = _list.count; lo < hi; ) {
        NSInteger mid = (lo + hi) / 2;
        NSInteger val = [_list[mid] integerValue];
        if (val == search) {
            return mid;
        } else if (val < search) {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    return NSNotFound;
}
- (NSInteger)middle {
    return _list.count / 2;
}
@end

File: swift/binary-search/Sources/BinarySearch/BinarySearch.swift
//Solution goes in Sources

enum BinarySearchError: Error {

    case unsorted
}

class BinarySearch {

    public var list: [Int]

    public var middle: Int

    public init(_ ary: [Int]) throws {

        if ary != ary.sorted() {
            throw BinarySearchError.unsorted
        }

        list = ary

        middle = list.count / 2


    }

    public func searchFor(_ search: Int) -> Int? {

        if list.isEmpty {
            return nil
        }

        var left = 0
        var right = list.count - 1

        while true {

            if left > right {
                return nil
            }

            let mid = (left + right) / 2

            if search == list[mid] {
                return mid
            } else if search < list[mid] {
                right = mid - 1
            } else {
                left = mid + 1
            }

        }

    }
}
